### PR TITLE
New Examples Transform Command

### PR DESCRIPTION
**Pull request by StarKhan6368**
**What**: Examples Transform Command

**Why**: The examples transform command enables you to transform your examples based on the new OpenAPI specification. This specification is generated by applying the provided overlay to the original specification.

**Checklist**:

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

### CODE REVIEW

File: core/src/main/kotlin/io/specmatic/core/pattern/JSONObjectPattern.kt [hunk 1]
```diff
@@ -27,6 +27,19 @@ data class JSONObjectPattern(
     val minProperties: Int? = null,
     val maxProperties: Int? = null
 ) : Pattern {
+
+    override fun eliminateOptionalKey(value: Value, resolver: Resolver): Value {
+        if (value !is JSONObjectValue) return value
+
+        val filteredPattern = this.pattern.filterKeys { !it.endsWith("?") }
```
Comment (joelrosario) [inline core/src/main/kotlin/io/specmatic/core/pattern/JSONObjectPattern.kt]: filteredPattern and filteredJsonObject variable names should be renamed for clarity. Perhaps mandatoryObjectPatternMap and mandatoryObjectMap.

File: core/src/main/kotlin/io/specmatic/core/pattern/JSONArrayPattern.kt [hunk 2]
```diff
@@ -13,6 +13,11 @@ import io.specmatic.core.value.Value
 import java.util.*
 
 data class JSONArrayPattern(override val pattern: List<Pattern> = emptyList(), override val typeAlias: String? = null) : Pattern, SequenceType {
+
+    override fun eliminateOptionalKey(value: Value, resolver: Resolver): Value {
+        return JSONArrayValue(emptyList())
```
Comment (joelrosario) [inline core/src/main/kotlin/io/specmatic/core/pattern/JSONArrayPattern.kt]: Either we should fully implement, or not at all. Given this is a legacy pattern, I suggest not implementing for now.

File: core/src/main/kotlin/io/specmatic/core/value/JSONObjectValue.kt [hunk 3]
```diff
@@ -86,6 +86,21 @@ data class JSONObjectValue(val jsonObject: Map<String, Value> = emptyMap()) : Va
 
     fun findFirstChildByName(name: String): Value? =
         jsonObject[name]
+
+    fun objectWithMandatoryKeys(
```
Comment (joelrosario) [inline core/src/main/kotlin/io/specmatic/core/value/JSONObjectValue.kt]: Do we still need this method? I believe it's not being called from anywhere.

**Review by joelrosario — CHANGES_REQUESTED** [review]

File: application/src/main/kotlin/application/ExamplesCommand.kt [hunk 4]
```diff
@@ -534,6 +538,36 @@ For example:
...
+        lateinit var contractFile: File
+
+        @Option(names = ["--overlay-file"], description = ["Overlay file path"], required = false)
+        val overlayFile: File? = null
+
+        @Option(names = ["--examples-dir"], description = ["Directory where existing examples reside"], required = true)
+        lateinit var examplesDir: File
+
+        @Option(names = ["--only-mandatory-keys-in-payload"], description = ["Transform existing examples so that they contain only mandatory keys in payload"], required = false)
+        var allowOnlyMandatoryKeysInPayload: Boolean = false
```
Comment (joelrosario) [inline application/src/main/kotlin/application/ExamplesCommand.kt]: this should be implicitly set to true if given, we should not have to say --only-mandatory-keys-in-payload=true

**Review by joelrosario — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/io/specmatic/core/examples/server/ExamplesInteractiveServer.kt [hunk 5]
```diff
@@ -724,6 +725,52 @@ class ExamplesInteractiveServer(
         private fun Map<String, String>.toValueMap(): Map<String, Value> {
             return this.mapValues { StringValue(it.value) }
         }
+
+        fun transformExistingExamples(contractFile: File, overlayFile: File?, examplesDir: File) {
+            val feature = parseContractFileToFeature(contractPath = contractFile.absolutePath, overlayContent = overlayFile?.readText().orEmpty())
+            val examples = examplesDir.getExamplesFromDir()
+
+            examples.forEach { example ->
+                consoleLog("\nTransforming ${example.file.nameWithoutExtension}")
```
Comment (joelrosario) [inline core/src/main/kotlin/io/specmatic/core/examples/server/ExamplesInteractiveServer.kt]: Put a verbose mode in, so we can switch the logging on if needed. There are too many logs if it's on by default.

File: core/src/main/kotlin/io/specmatic/core/pattern/JSONObjectPattern.kt (hunk 1 above)
Comment (StarKhan6368) [reply core/src/main/kotlin/io/specmatic/core/pattern/JSONObjectPattern.kt]: Thanks, I've fixed that

File: core/src/main/kotlin/io/specmatic/core/pattern/JSONArrayPattern.kt (hunk 2 above)
Comment (StarKhan6368) [reply core/src/main/kotlin/io/specmatic/core/pattern/JSONArrayPattern.kt]: Thanks, removed the implementation

File: core/src/main/kotlin/io/specmatic/core/value/JSONObjectValue.kt (hunk 3 above)
Comment (StarKhan6368) [reply core/src/main/kotlin/io/specmatic/core/value/JSONObjectValue.kt]: Agree, i've dropped the method now

File: application/src/main/kotlin/application/ExamplesCommand.kt (hunk 4 above)
Comment (StarKhan6368) [reply application/src/main/kotlin/application/ExamplesCommand.kt]: yeah makes sense, I've addressed this, also added a few simple tests for now.

**Review by joelrosario — APPROVED** [review]

**Pull request merged**